packages/contracts/src/PotLuck.sol

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/**
+ * @title PotLuck
+ * @notice A recurring pool of token contributions where one random participant wins the pot each period.
+ *         Creator configures entry amount, period, optional max participants, and public access.
+ *         Platform collects a fee percentage on each payout.
+ */
+contract PotLuck is Ownable {
+    using SafeERC20 for IERC20;
+
+    uint8 constant MAX_PARTICIPANTS = 100; // Max participants per pot
+
+    uint256 public potCount;
+    uint256 public platformFeePct; // e.g., 10 for 10%
+
+    struct Pot {
+        uint256    id;
+        uint32     round;         // current round number
+        uint256     deadline;      // timestamp when round ends
+        uint256    balance;      // total balance in pot
+        address    token;
+        uint256    entryAmount;
+        uint256    period;          // in seconds
+        bool       isPublic;
+        uint32     totalParticipants;
+        address[]  participants;   // list of participants in current round
+    }
+
+    mapping(uint256 => Pot) public pots;
+    mapping(bytes32 => bool) public hasJoinedRound; // keccak256(potId,roundId,user) -> bool;
+    mapping(bytes32 => bool) public hasWon; // keccak256(potId,user) -> bool;
+
+    event PotCreated(uint256 indexed potId, address indexed creator);
+    event Joined(uint256 indexed potId, uint32 roundId,address indexed user);
+    event Payout(uint256 indexed potId, address indexed winner, uint256 amount, uint32 round);
+
+    constructor(uint256 _platformFeePct) Ownable(msg.sender) {
+        require(_platformFeePct <= 100, "Fee pct <= 100");
+        platformFeePct = _platformFeePct;
+    }
+
+    /**
+     * @notice Create a new public pot with specified parameters.
+     * @param token Address of the ERC20 token contract.
+     * @param entryAmount USDC amount each participant must contribute.
+     * @param periodSeconds Duration of each round in seconds.
+     */
+    function createPotLuck(
+        address token,
+        uint256 entryAmount,
+        uint256 periodSeconds
+    ) external {
+        require(entryAmount > 0, "Entry > 0");
+        require(periodSeconds >= 1 hours, "Period must be >=1h");
+
+        uint256 currentPotCount = potCount++;
+        Pot storage p = pots[currentPotCount];
+        p.id = currentPotCount;
+        p.token = token;
+        p.entryAmount = entryAmount;
+        p.period = periodSeconds;
+        p.isPublic = true;
+        p.totalParticipants = 1;
+        p.deadline = block.timestamp + periodSeconds;
+        p.participants.push(msg.sender);
+
+        IERC20(token).safeTransferFrom(msg.sender, address(this), entryAmount);
```

Comment:
Platform fee should only be charged at potluck creation. So we need an additional transfer to our treasury address, which should be set when contract is created.